### ISSUE TITLE
Matt: accumulated Fl_Preferences issues, are there more?

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#303]
I will be fixing all complaints about Fl_Preferences soon. 

This Issue is filed to make sure that I sum up everything correctly. Comments are welcome. 

I won't fix Bill's complaint though that file formats have no business in a GUI library and Fl_Prefs should be removed. Even though he's probably right ;-) 🙋🏻‍♂️.

 - ✔︎ [https://github.com/fltk/fltk/issues/227](https://github.com/fltk/fltk/issues/227) Prefs path without creating file, also an 'exists()' method
 - ✔︎ [https://www.fltk.org/str.php?L3235](https://www.fltk.org/str.php?L3235) A crash bug including a diff file to fix it
 - ✔︎ [Forum](https://www.fltk.org/newsgroups.php?s1+gfltk.coredev+v19883) Documentation
 - ✔︎ [Forum](https://www.fltk.org/newsgroups.php?s1+gfltk.coredev+v18948+T0) Path generation (may already be fixed)
 - ✔︎ [Forum](https://www.fltk.org/newsgroups.php?s1+gfltk.coredev+v19565+T0) Issues with different "locale" settings
 - won't fix until I have better suggestions: [Forum](https://www.fltk.org/newsgroups.php?s1+gfltk.general+v35753+T0) Better Unix file location than ``~/.fltk/``
 - won't fix: entries that wrap at line length are not Unicode UTF-8 aware
 - won't fix: More documentation (link to INI file Wikipedia, explain intentions and usage, examples, maybe tests)
 - won't fix: Maybe: adhere better to the .ini file "standard" without breaking things
 - won't fix: Maybe: manage writes from multiple apps (merge data? Read-modify-write cycle?)

**Comment by Albrecht-S** [comment]
Note: I edited a link in the comment above to make it accessible for non-devs.

Issue #79 maybe: "Fltk doesn't respect XDG Base Directory standard" (currently assigned to me, please feel free to assign to yourself).

Issue 79 mentions [STR 3370](https://www.fltk.org/str.php?L3370): "use freedesktop.org XDG basedir-spec" (also assigned to me, feel free to take over).

**Comment by Albrecht-S** [comment]
> Maybe: adhere better to the .ini file "standard" without breaking things

I don't think that this is useful, I'd set the priority of this one very low. IMHO it's OK to mention that the syntax is similar to the "ini file standard" but nothing else is required. YMMV.

**Comment by MatthiasWM** [comment]
OK, I think I caught everything pressing in PR #374. I decided against adding support for the XDG based standard simply because I don't have a Linux machine available ATM. Also, if would break existing apps. I am open for a discussion, of course.

**Comment by Albrecht-S** [comment]
> I decided against adding support for the XDG based standard simply because I don't have a Linux machine available ATM.

I could probably assist at least in testing.

> Also, if would break existing apps. I am open for a discussion, of course.

I've personally "seen" such a transition in Git config files. If you like to see what happens, `sctrace` can be used to trace the Git invocation. It does basically check "both" locations and uses the one that can be accessed:
```
$ strace git config -l 2>&1 | grep /home/albrecht | grep config
access("/home/albrecht/.config/git/config", R_OK) = 0
...
access("/home/albrecht/.gitconfig", R_OK) = -1 ENOENT (No such file or directory)
```

In my case I moved the "classic" config file `/home/albrecht/.gitconfig` to the new `XDG` config file `/home/albrecht/.config/git/config` and Git happily opens the latter.

IMHO something like this could be done in FLTK 1.4 as well. Please keep in mind that you did already change the preferences file location in partially incompatible ways WRT FLTK 1.3.x.

I'd appreciate if we could do a further change to respect the `XDG*` standard on Linux and I'd be willing to assist.

FTR: Issue #79 is still open ...

**Comment by MatthiasWM** [comment]
Ok, I can give that a try. It does not look too complicated to implement. So does the pure existence of `/home/albrecht/.config/` imply that XDG is supported? Or should we always write new preference there, even if the folder was not created yet? Hmm, guess I actually have to read the docs 😉

**Comment by Albrecht-S** [comment]
Reading existing preferences is simple, we just need to take what we get. If there are both files (i.e. the same preferences file in both directories) I don't know. There must be a preference (pun not intended) but I'd suggest not to try to merge different files.

Citing the docs mentioned above: "_There is a single base directory relative to which user-specific configuration files should be written. This directory is defined by the environment variable $XDG_CONFIG_HOME._"

IIRC there's also a fallback but I don't know.

For writing preferences a quick thought is (with this order):
1. if a particular preferences file already exists, use it
2. if an FLTK preferences directory exists (but not 1.), create a new file in this directory
3. if none exists, use `$XDG_CONFIG_HOME` (or whatever the correct environment variable is) and create FLTK config files below that dir
4. if `$XDG_CONFIG_HOME` is not defined, use the "classic" scheme

... or something like that.

We should never (try to) move FLTK config files from one folder to another. This is something the user must do if they want to move their preferences files from `$HOME/.fltk` to `$XDG_CONFIG_HOME/fltk`.

My take on this is: the FLTK config directory is either `$HOME/.fltk` or `$XDG_CONFIG_HOME/fltk`, the preferences files below that config dir should be in the same directory structure as before such that users can do:
```
$ mv $HOME/.fltk $XDG_CONFIG_HOME/fltk
```
to move all their FLTK preferences files from their home directory to the new config directory. This could be documented and FLTK could work as before.

**Comment by Albrecht-S** [comment]
@MatthiasWM Some more info:
> So does the pure existence of /home/albrecht/.config/ imply that XDG is supported?

I think that we should always test for more than one option. IMHO it's most important to find existing files so users don't lose their preferences.

On my system (i.e. for me as user albrecht) `$XDG_CONFIG_HOME` is not defined at all but nevertheless there are lots of files (11) and subdirectories (58) in `~/.config`. The spec says:

> $XDG_CONFIG_HOME defines the base directory relative to which user-specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.

The latter seems to be the case on my system and this is why Git works as described above and why all the other configs are stored in `$HOME/.config` (`/home/albrecht/.config`). But there are still lots of other config files in the old `.<application>` config dirs.

There are similar defaults for system config files IIRC, but that's another (although related) problem.

I think if we can make sure that we don't ignore existing files we should always create new preferences files in the "new" (XDG) directories even if `$XDG_CONFIG_HOME` is not defined (see default definition above). The next time we try to read this file we'd find it.

The only problem with this strategy would be if a user uses both FLTK 1.3 and FLTK 1.4 on the same system. FLTK 1.3 would only find the "old" location (even if a file exists in the "new" location) and FLTK 1.4 would create new files in the "new" location if an old file does not exist. This could be confusing, for instance: run fluid (1.4) on a new system, this creates `$XDG_CONFIG_HOME/...` preferences. Then run fluid (1.3) which would create `$HOME/.fltk/...` preferences - which ends up having both preferences files.

**Comment by MatthiasWM** [comment]
As I am reading myself into XDG a bit more, `$XDG_CONFIG_HOME` can be set to any shell-readable path, including `~` and recursive `$` environment variables, and even multiple paths separated by `:`'s. Not sure if there is an API to interpret those things, or if it needs to be written from scratch?!

**Comment by Albrecht-S** [comment]
Off the top of my head: as I understand `$XDG_CONFIG_HOME` should be one path. I don't know if a path can be constructed from recursive environment variables, but maybe yes. However ISTR that all such paths must be absolute paths. If this is all possible it looks pretty complicated to parse and translate everything. Does `~` count as an absolute path? Can the user's `PATH` environment variable contain other environment variables, or must such variables be resolved when setting the `PATH`? Questions I can't answer but I'd assume that only real paths can be part of the `PATH` variable.

Multiple paths can indeed be defined, but IIRC only in the additional environment variable `$XDG_CONFIG_DIRS` - which doesn't make it much easier though.

I've just skimmed over the specs again. The only rule I found was _**"All paths set in these environment variables must be absolute"**_. To me this means that I can use such paths in a file open statement w/o further interpretation of env vars. The same rule would apply for the colon-separated paths in `$XDG_CONFIG_DIRS`.

BTW: if any of these path variables could be recursive environment variables this would IMHO open a big security hole. I don't think that this is intended or allowed.

**Comment by Albrecht-S** [comment]
After thinking further I do also believe that `~` is only a shell convention (to be interpreted like `$HOME`) which in turn would mean that this is not allowed in a **path**.

**Comment by Albrecht-S** [comment]
@MatthiasWM Are we reading the same document? In [my specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) the word "shell" does not exist.

**Comment by MatthiasWM** [comment]
@Albrecht-S Thanks again for the link. I had been collecting general use cases from stack overflow, but reading through the specs again, well, those cases are simply wrong. So the strategy is now: check if a prefs file or data directory exists in `~/.fltk/Vendor/Application`, we will use that for everything. If not, we use `$XDG_DATA_HOME/Vendor/Application/` (not `fltk`) for user data files, and `$XDG_CONFIG_HOME/Vendor/Application.prefs` are used for preferences.

As a precaution, I check for `~/`, `$HOME/`, and `${HOME}/` at the start and replace them. I'll truncate at a possible `:`. If either of the environment variables are not set, I will use their respective default. 

Porting would be done calling
```
mkdir ~/.local/share/Vendor/
mv ~/.fltk/Vendor/Application/ ~/.local/share/Vendor/
mv ~/.fltk/Vendor/ ~/.config/
```

Did I catch everything?

**Comment by MatthiasWM** [comment]
Another question. Back then, I used CamelCode to write the functions. Should this be corrected:
```
  char group_exists( const char *key );
  // Deprecated
  char groupExists( const char *key ) { return group_exists(key); }
  char delete_group( const char *group );
  // Deprecated
  char deleteGroup( const char *group ) { return delete_group(group); }
```
It's not too many functions, but code that uses the new methods will not compile in FLTK 1.3.x .

**Comment by Albrecht-S** [comment]
> So the strategy is now: check if a prefs file or data directory exists in ~/.fltk/Vendor/Application, we will use that for everything. If not, we use $XDG_DATA_HOME/Vendor/Application/ (not fltk) for user data files, and $XDG_CONFIG_HOME/Vendor/Application.prefs are used for preferences.

Yes, but... If one of the "vendors" is `"fltk(.org)"` so we can use `$XDG_CONFIG_HOME/fltk.org/fluid.prefs` etc. But then, the "standard" appears to be to use "application" as the primary key in `$XDG_CONFIG_HOME`, hence `fltk.org` would look a little strange. I have such subdirectories like Code, brasero, vlc, VirtualBox, xviewer, pulse, google-chrome, gnome-session to name only a few.

On a first look `fltk` would be fine, but then ... other applications/vendors might want to have their own top level directory in `$XDG_CONFIG_HOME`. I don't know.

> Porting would be done calling ... Did I catch everything?

No, I don't think what you wrote is correct, but we might have a "problem" here. The spec separates "config" (i.e. preferences) in `$XDG_CONFIG_HOME` and "data" in `$XDG_DATA_HOME`. I believe that **preference files** belong in `$XDG_CONFIG_HOME` which defaults to `$HOME/.config` (not `.local/share`). According to the spec **data** files belong in `$XDG_DATA_HOME` which defaults to `$HOME/.local/share/`. I would think that our additional FLTK data directory which is currently side by side with the preference file **could** be placed in `$XDG_DATA_HOME` which would require a split of preference aka "config" files and "data" files. I would not recommend to do this and rather leave the structure as it is, i.e. preference file and additional data directory side by side, even if this means that everything is in `$XDG_CONFIG_HOME`.

But I may have misunderstood or misinterpreted something. In my "simple" model porting would only require one directory to be moved but ... it's maybe not that simple.

**Comment by Albrecht-S** [comment]
> Back then, I used CamelCode to write the functions. Should this be corrected: ...

My answer is yes, change this to FLTK standard with deprecated "old" names.

> ... but code that uses the new methods will not compile in FLTK 1.3.x

That doesn't matter. As long as the old name(s) still exist users that target both 1.3 and 1.4 can use the old name. The other way around is not necessary. It's the same with our new name FL/platform.H which replaces FL/x.H (old is good for both, new only for 1.4).

**Comment by MatthiasWM** [comment]
@Albrecht-S : Hmm, as for the Vendor/Application naming, it's kind of arbitrary anyway. Users can read how the filename is created and set Vendor and Application accordingly. I find the XDG approach too simple, Fluid, for example, is also the name of a web browser, and for several fluid simulations. 

I'll be happy to put the user data into the config directory. Less work for me. And only one `mv` to copy all preferences and data into the new system.

**Comment by Albrecht-S** [comment]
@MatthiasWM FYI: one of the recent commits must have introduced new doxygen warnings:
```
[2/2] Generating HTML documentation
/git/fltk/master/FL/Fl_Preferences.H:375: warning: reached end of file while inside a '```' block!
The command that should end the block seems to be missing!

/git/fltk/master/src/Fl_Preferences.cxx:1889: warning: reached end of file while inside a '```' block!
The command that should end the block seems to be missing!

/git/fltk/master/src/Fl_Preferences.cxx:32: warning: documented symbol 'char Fl_Preferences::nameBuffer' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:33: warning: documented symbol 'char Fl_Preferences::uuidBuffer' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:34: warning: documented symbol 'Fl_Preferences * Fl_Preferences::runtimePrefs' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:35: warning: documented symbol 'unsigned int Fl_Preferences::fileAccess_' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:66: warning: documented symbol 'const char * Fl_Preferences::newUUID' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:106: warning: documented symbol 'void Fl_Preferences::file_access' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:116: warning: documented symbol 'unsigned int Fl_Preferences::file_access' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:141: warning: no uniquely matching class member found for 
  Fl_Preferences::Root Fl_Preferences::filename(char *buffer, size_t buffer_size, Root root, const char *vendor, const char *application)

/git/fltk/master/src/Fl_Preferences.cxx:225: warning: documented symbol 'Fl_Preferences::Fl_Preferences' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:242: warning: documented symbol 'Fl_Preferences::Fl_Preferences' was not declared or defined.
/git/fltk/master/src/Fl_Preferences.cxx:258: warning: documented symbol 'Fl_Preferences::Fl_Preferences' was not declared or defined.
```

**Comment by MatthiasWM** [comment]
Yes, that should be fixed in my branch. I will push it upstream as soon as I have XDG implemented.

**Comment by MatthiasWM** [comment]
Fixed in 7308bcdb74e34626c6459699cb57371afd7b343b
